BoC search changes/improvements

Buurazu opened the pull request:
I've made some changes to the search function's Enter input based on how I think it would control best.
I would like feedback from @CreepSore and anyone else about it, especially as I think he may have had something planned with EID.BoCSLockMode which currently is only 0.

Changes:
When not in search mode: Pressing Enter will both wipe your previous search, and start a new one
(In the future there can be two configurable hotkeys bound for these; for example Backspace to clear search, but Enter to start a new one / modify your previous one)
When in search mode: Pressing Enter will save your current search
Also, unlocking no longer wipes your current search; I think you may want to keep your search when unlocking

General improvements:
Do the item name check once per item ID rather than once per recipe
Support Shift in the search input and disable regex (so ?, !, and . all work now)
Check for MCM being open
Sanitize accent marks out of item names
Check English names as well as the localized language (mostly for korean/chinese/japanese; could have a hardcoded list of languages to check english on, if other languages don't like items appearing in their searches based on english name)
Pressing Escape mid-search will cancel a search
Shorten backspace delay to more match general Windows backspace delay

CreepSore commented:
>I think he may have had something planned with EID.BoCSLockMode which currently is only 0.

Generally, the `EID.BoCSLockMode ` Parameter is planned exactly for this type of handling.
If possible try to make an own mode for that.
I already thought about making the functions `eid_bagofcrafting_search.lua` a kind of wrapper for `eid_bagofcrafting_<MODE>.lua` files - so we have a "faked" object oriented pattern.
This would make implementing and maintaining a lot easier since you have a singular file for every type of LockMode.
I could do that though if you want.

>Pressing Enter will save your current search

This was explicitly not wished as default behavior as stated in https://github.com/wofsauge/External-Item-Descriptions/issues/660#issuecomment-2142078011

I did say however that we could make multiple locking modes, that's why we have the parameter.

>Support Shift in the search input 

Haha i actually thought about that but left it as is since the filtering itself is case insensitive.
It does have a little bit of a nicer feeling with this solution.

>Check for MCM being open

Did you manage to open MCM while in search mode?
This should actually get blocked completely by https://github.com/wofsauge/External-Item-Descriptions/blob/05e5e2db61538998736fb9e1100010b5bc57facd/eid_bagofcrafting_search.lua#L317-L322

Also when trying your branch I get following error (tried 2 times):
`[INFO] - Lua Debug: Error in "MC_HUD_RENDER" call from "External Item Descriptions": ...th/mods/external-item-descriptions/eid_bagofcrafting.lua:1219: attempt to index a nil value
Stack Traceback:
  ...th/mods/external-item-descriptions/eid_bagofcrafting.lua:1219: in function 'handleBagOfCraftingRendering'
  ...f Isaac Rebirth/mods/external-item-descriptions/main.lua:1358: in function at line 1220`

Buurazu commented:
The MCM issue was with initiating a search (by pressing Enter to go into a mod's options) while MCM is open and having all controls locked. Discovered it while trying to change language.

Looks like the error is because I changed to getting the item name with a function that can get English but apparently doesn't work for modded items, I was testing with vanilla only. The API functions aren't great when it comes to ignoring language choice, for the time being I just added a new function to get the english name of vanilla+modded items. Sorry about that.

I didn't see wof's comment but he did mention playtesting and I definitely personally think users would expect to keep the search term upon pressing enter; how are they going to gather what they need for the recipe if it goes away? So whatever may happen with settings, I would strongly vote for keeping search usable solely with Enter, with it clearing+starting search and saving+finishing search by default.

I'm very impressed with your work on this and if you have intents to clean up / compartmentalize other parts of the code that would be great, BoC improvement is where I got started working on EID (refreshing, locking, the asynchronous recipe crunching is all me) and it's probably a bit of a mess at this point...

CreepSore commented:
>I didn't see wof's comment but he did mention playtesting and I definitely personally think users would expect to keep the search term upon pressing enter; how are they going to gather what they need for the recipe if it goes away?

Yeah, I also think it's better keeping the search open since the main reason one wants to search is because of the recipe.
I've played 2 runs and i never needed the search option without locking afterwards.

>'m very impressed with your work on this and if you have intents to clean up / compartmentalize other parts of the code that would be great, BoC improvement is where I got started working on EID (refreshing, locking, the asynchronous recipe crunching is all me) and it's probably a bit of a mess at this point...

Thanks :)
Insane work from your part too!

I think the BoC part would profit from making multiple modules.
Given that require only "appends" the files we won't lose any performance anyway, so we can only win from refactoring.
I'll start refactoring my part for now (since it's independent) and after that is done we can think about refactoring the main BoC module into smaller parts.

I think at least the callbacks, the crunching function and the util functions could go in an own file each.

wofsauge commented:
> I've played 2 runs and i never needed the search option without locking afterwards.

Then lets do the version where the search is not cleared after closing the search feature. My only concern for that was, that people forget that they have a search query and will then complain that no recipes are shown.
